flocker_bb/ec2_buildslave.py

```diff
@@ -13,436 +13,152 @@
-        #
-        # We currently discard the keypair data because we don't need it.
-        # If we do need it in the future, we will always recreate the keypairs
-        # because there is no way to
-        # programmatically retrieve the private key component, unless we
-        # generate it and store it on the filesystem, which is an unnecessary
-        # usage requirement.
-        try:
-            key_pair = self.conn.get_all_key_pairs(keypair_name)[0]
-            assert key_pair
-            # key_pair.delete() # would be used to recreate
-        except boto.exception.EC2ResponseError, e:
-            if 'InvalidKeyPair.NotFound' not in e.body:
-                if 'AuthFailure' in e.body:
-                    print ('POSSIBLE CAUSES OF ERROR:\n'
-                           '  Did you sign up for EC2?\n'
-                           '  Did you put a credit card number in your AWS '
-                           'account?\n'
-                           'Please doublecheck before reporting a problem.\n')
-                raise
-            # make one; we would always do this, and stash the result, if we
-            # needed the key (for instance, to SSH to the box).  We'd then
-            # use paramiko to use the key to connect.
-            self.conn.create_key_pair(keypair_name)
-
-        # create security group
-        try:
-            group = self.conn.get_all_security_groups(security_name)[0]
-            assert group
-        except boto.exception.EC2ResponseError, e:
-            if 'InvalidGroup.NotFound' in e.body:
-                self.security_group = self.conn.create_security_group(
-                    security_name,
-                    'Authorization to access the buildbot instance.')
-                # Authorize the master as necessary
-                # TODO this is where we'd open the hole to do the reverse pb
-                # connect to the buildbot
-                # ip = urllib.urlopen(
-                #     'http://checkip.amazonaws.com').read().strip()
-                # self.security_group.authorize('tcp', 22, 22, '%s/32' % ip)
-                # self.security_group.authorize('tcp', 80, 80, '%s/32' % ip)
+        self.building = set()
+
+        self.ec2 = EC2(
+            access_key=identifier,
+            secret_access_token=secret_identifier,
+            region=region,
+            name=name,
+            image_name=image_name,
+            size=instance_type,
+            keyname=keypair_name,
+            security_groups=[security_name],
+            userdata=user_data,
+            metadata=tags,
+        )
+
+        self.addService(TimerService(60, self.periodic))
+
+    def buildStarted(self, sb):
+        self._clearBuildWaitTimer()
+        self.building.add(sb.builder_name)
+
+    def buildFinished(self, sb):
+        BuildSlave.buildFinished(self, sb)
+
+        self.building.discard(sb.builder_name)
+        if not self.building:
+            if self.build_wait_timeout == 0:
```

Comment:
The difference between `==` here and `<=` in `_setBuildWaitTimer` below seems fishy.